src/ring_theory/dedekind_domain.lean

```diff
@@ -108,33 +108,148 @@ structure is_dedekind_domain_dvr : Prop :=
 (is_dvr_at_nonzero_prime : ∀ P ≠ (⊥ : ideal A), P.is_prime →
   discrete_valuation_ring (localization.at_prime P))
 
+section inverse
+
+open_locale classical
+
+variables {R₁ : Type*} [integral_domain R₁] {g : fraction_map R₁ K}
+
+
+variables {I J : fractional_ideal g}
+
+noncomputable instance : has_inv (fractional_ideal g) := ⟨λ I, 1 / I⟩
+
+lemma inv_eq : I⁻¹ = 1 / I := rfl
+
+lemma inv_zero' : (0 : fractional_ideal g)⁻¹ = 0 := fractional_ideal.div_zero
+
+lemma inv_nonzero {J : fractional_ideal g} (h : J ≠ 0) :
+J⁻¹ = ⟨(1 : fractional_ideal g) / J, fractional_ideal.fractional_div_of_nonzero h⟩ :=
+fractional_ideal.div_nonzero _
+
+lemma coe_inv_of_nonzero {J : fractional_ideal g} (h : J ≠ 0) :
+  (↑J⁻¹ : submodule R₁ g.codomain) = g.coe_submodule 1 / J :=
+by { rwa inv_nonzero _, refl, assumption}
+
+/-- `I⁻¹` is the inverse of `I` if `I` has an inverse. -/
+theorem right_inverse_eq (I J : fractional_ideal g) (h : I * J = 1) :
+  J = I⁻¹ :=
+begin
+  have hI : I ≠ 0 := fractional_ideal.ne_zero_of_mul_eq_one I J h,
+  suffices h' : I * (1 / I) = 1,
+  { exact (congr_arg units.inv $
+      @units.ext _ _ (units.mk_of_mul_eq_one _ _ h) (units.mk_of_mul_eq_one _ _ h') rfl) },
+  apply le_antisymm,
+  { apply fractional_ideal.mul_le.mpr _,
+    intros x hx y hy,
+    rw mul_comm,
+    exact (fractional_ideal.mem_div_iff_of_nonzero hI).mp hy x hx },
+  rw ← h,
+  apply fractional_ideal.mul_left_mono I,
+  apply (fractional_ideal.le_div_iff_of_nonzero hI).mpr _,
+  intros y hy x hx,
+  rw mul_comm,
+  exact fractional_ideal.mul_mem_mul hx hy
+end
+
+theorem mul_inv_cancel_iff {I : fractional_ideal g} :
+  I * I⁻¹ = 1 ↔ ∃ J, I * J = 1 :=
+⟨λ h, ⟨I⁻¹, h⟩, λ ⟨J, hJ⟩, by rwa [← @right_inverse_eq _ _ _ _ _ I J hJ]⟩
+
+variables {K' : Type*} [field K'] {g' : fraction_map R₁ K'}
+
+@[simp] lemma map_inv (I : fractional_ideal g) (h : g.codomain ≃ₐ[R₁] g'.codomain) :
+  (I⁻¹).map (h : g.codomain →ₐ[R₁] g'.codomain) = (I.map h)⁻¹ :=
+by rw [inv_eq, fractional_ideal.map_div, fractional_ideal.map_one, inv_eq]
+
+open_locale classical
+
+open submodule submodule.is_principal
+
+@[simp] lemma span_singleton_inv (x : g.codomain) :
+  (fractional_ideal.span_singleton x)⁻¹ = fractional_ideal.span_singleton (x⁻¹) :=
+fractional_ideal.one_div_span_singleton x
+
+local attribute [semireducible] fractional_ideal.span_singleton
```

Comment:
Should this be documented?

Reply:
I guess it's a good idea to have a commented explaining why we want this.
Should it be an `abbreviation` globally?

Reply:
Apologies for the naive question, but what is an `abbreviation`? Also, I guess @bryangingechen say "this" to mean the fact that `[semireducible]` is required, right?

Reply:
Yes, that's what I meant.

Reply:
I have removed it directly: it was not needed anymore. Thanks for catching this.